File: builds/java/eclipse-e4/org.eclipse.e4.javascript/src/org/eclipse/e4/internal/javascript/JSImport.java
/*******************************************************************************
 * Copyright (c) 2009 IBM Corporation and others
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package org.eclipse.e4.internal.javascript;

import org.eclipse.e4.javascript.Constants;

import java.util.*;
import java.util.Map.Entry;

public class JSImport {
	private String name;
	private VersionRange versionRange = VersionRange.emptyRange;
	private String bundleSymbolicName;
	private VersionRange bundleVersionRange = VersionRange.emptyRange;
	private boolean optional = false;
	private Map attributes = new HashMap();
	private Map directives = new HashMap();
	private JSExport wiredExport;

	public JSImport(String header) {
		if (header == null)
			throw new IllegalArgumentException("header cannot be null"); //$NON-NLS-1$
		parseImport(header);
	}

	private void parseImport(String header) {
		StringTokenizer tokenizer = new StringTokenizer(header, Constants.PARAMETER_DELIMITER);
		this.name = tokenizer.nextToken().trim();
		while (tokenizer.hasMoreTokens()) {
			String token = (String) tokenizer.nextElement();
			if (token.indexOf(Constants.DIRECTIVE_EQUALS) != -1)
				parseDirective(token);
			else if (token.indexOf(Constants.ATTRIBUTE_EQUALS) != -1)
				parseAttribute(token);
			else
				throw new IllegalArgumentException("bad import syntax: " + token + " in " + header); //$NON-NLS-1$//$NON-NLS-2$
		}
	}

	private void parseAttribute(String token) {
		int index = token.indexOf(Constants.ATTRIBUTE_EQUALS);
		String attributeName = token.substring(0, index).trim();
		if (attributeName.length() == 0)
			return;

		Object value = token.substring(index + Constants.ATTRIBUTE_EQUALS.length()).trim();

		if (attributeName.equals(Constants.VERSION_ATTRIBUTE))
			versionRange = new VersionRange((String) value);
		else if (attributeName.equals(Constants.BUNDLE_SYMBOLIC_NAME_ATTRIBUTE))
			bundleSymbolicName = (String) value;
		else if (attributeName.equals(Constants.BUNDLE_VERSION_ATTRIBUTE))
			bundleVersionRange = new VersionRange((String) value);
		attributes.put(attributeName, value);
	}

	private void parseDirective(String token) {
		int index = token.indexOf(Constants.DIRECTIVE_EQUALS);
		String directiveName = token.substring(0, index).trim();
		if (directiveName.length() == 0)
			return;

		String value = token.substring(index + Constants.DIRECTIVE_EQUALS.length()).trim();
		if (directiveName.equals(Constants.RESOLUTION_DIRECTIVE))
			optional = Constants.RESOLUTION_OPTIONAL.equals(value);
		directives.put(directiveName, value);
	}

	public String getName() {
		return name;
	}

	public VersionRange getVersionRange() {
		return versionRange;
	}

	public String getBundleSymbolicName() {
		return bundleSymbolicName;
	}

	public VersionRange getBundleVersionRange() {
		return bundleVersionRange;
	}

	public boolean isOptional() {
		return optional;
	}

	public Map getAttributes() {
		return attributes;
	}

	public Map getDirectives() {
		return directives;
	}

	public boolean wire(JSExport candidate) {
		if (!name.equals(candidate.getName()))
			return false;

		for (Iterator iterator = attributes.entrySet().iterator(); iterator.hasNext();) {
			Entry entry = (Entry) iterator.next();
			String key = (String) entry.getKey();

			if (key.equals(Constants.VERSION_ATTRIBUTE)) {
				if (!versionRange.isIncluded(candidate.getVersion()))
					return false;
			} else if (key.equals(Constants.BUNDLE_SYMBOLIC_NAME_ATTRIBUTE)) {
				if (!bundleSymbolicName.equals(candidate.getBundleSymbolicName()))
					return false;
			} else if (key.equals(Constants.BUNDLE_VERSION_ATTRIBUTE)) {
				if (!bundleVersionRange.isIncluded(candidate.getBundleVersion()))
					return false;
			} else {
				Object value = entry.getValue();
				Object attributeValue = candidate.getAttributes().get(key);
				if (attributeValue == null || !attributeValue.equals(value))
					return false;
			}
		}

		for (Iterator iterator = candidate.getMandatory().iterator(); iterator.hasNext();) {
			String key = (String) iterator.next();
			Object mandatoryValue = candidate.getAttributes().get(key);
			Object value = attributes.get(key);
			if (value == null || !value.equals(mandatoryValue))
				return false;
		}

		wiredExport = candidate;
		return true;
	}

	public JSExport getWiredExport() {
		return wiredExport;
	}

	public void unwire() {
		wiredExport = null;
	}

}

File: builds/java/eclipse-e4/org.eclipse.e4.javascript/scripts/JSImport.js
function JSImport(header) {
	if (header === null)
		throw "header cannot be null";

	this._name = null;
	this._versionRange = VersionRange.EMPTY_RANGE;
	this._bundleSymbolicName = null;
	this._bundleVersionRange = VersionRange.EMPTY_RANGE;
	this._optional = false;
	this._attributes = {};
	this._directives = {};
	this._wiredExport = null;

	_parseImport(header);
}

JSImport.prototype = {
	_parseImport : function(header) {
		var tokens = header.split(Constants.PARAMETER_DELIMITER);
		this._name = tokens[0].replace(/^\s+|\s+$/g, '');
		for ( var i = 1; i < tokens.length; i++) {
			var token = tokens[i];
			if (token.indexOf(Constants.DIRECTIVE_EQUALS) !== -1)
				this._parseDirective(token);
			else if (token.indexOf(Constants.ATTRIBUTE_EQUALS) !== -1)
				this._parseAttribute(token);
			else
				throw "bad import syntax: " + token + " in " + header;
		}
	},
	_parseAttribute : function(token) {
		var index = token.indexOf(Constants.ATTRIBUTE_EQUALS);
		var attributeName = token.substring(0, index).replace(/^\s+|\s+$/g, '');
		if (attributeName.length === 0)
			return;

		var value = token.substring(index + Constants.ATTRIBUTE_EQUALS.length).replace(/^\s+|\s+$/g, '');

		if (attributeName === Constants.VERSION_ATTRIBUTE)
			this._versionRange = VersionRange.parseVersionRange(value);
		else if (attributeName === Constants.BUNDLE_SYMBOLIC_NAME_ATTRIBUTE)
			this._bundleSymbolicName = value;
		else if (attributeName === Constants.BUNDLE_VERSION_ATTRIBUTE)
			this._bundleVersionRange = VersionRange.parseVersionRange(value);
		this._attributes[attributeName] = value;
	},
	_parseDirective : function(token) {
		var index = token.indexOf(Constants.DIRECTIVE_EQUALS);
		var directiveName = token.substring(0, index).replace(/^\s+|\s+$/g, '');
		if (directiveName.length === 0)
			return;

		var value = token.substring(index + Constants.DIRECTIVE_EQUALS.length).replace(/^\s+|\s+$/g, '');
		if (directiveName === Constants.RESOLUTION_DIRECTIVE && value === Constants.RESOLUTION_OPTIONAL)
			this._optional = true;
		this._directives[directiveName] = value;
	},
	getName : function() {
		return this._name;
	},
	getVersionRange : function() {
		return this._versionRange;
	},
	getBundleSymbolicName : function() {
		return this._bundleSymbolicName;
	},
	getBundleVersionRange : function() {
		return this._bundleVersionRange;
	},
	isOptional : function() {
		return this._optional;
	},
	getAttributes : function() {
		this._attributes;
	},
	getDirectives : function() {
		return this._directives;
	},
	wire : function(candidate) {
		if (this._name !== candidate.getName())
			return false;

		if (!this._checkAttributes(candidate))
			return false;

		if (!this._checkMandatory(candidate))
			return false;

		this._wiredExport = candidate;
		return true;
	},
	_checkAttributes : function(candidate) {
		for ( var key in this._attributes) {
			if (key === Constants.VERSION_ATTRIBUTE) {
				if (!this._versionRange.isIncluded(candidate.getVersion()))
					return false;
			} else if (key === Constants.BUNDLE_SYMBOLIC_NAME_ATTRIBUTE) {
				if (!this._bundleSymbolicName === candidate.getBundleSymbolicName())
					return false;
			} else if (key === Constants.BUNDLE_VERSION_ATTRIBUTE) {
				if (!this._bundleVersionRange.isIncluded(candidate.getBundleVersion()))
					return false;
			} else {
				var value = this._attributes[key];
				var attributeValue = candidate.getAttributes()[key];
				if (attributeValue !== value)
					return false;
			}
		}
		return true;
	},
	_checkMandatory : function(candidate) {
		for ( var i = 0; i < candidate.getMandatory().length; i++) {
			var key = candidate.getMandatory()[i];
			var mandatoryValue = candidate.getAttributes()[key];
			var value = this._attributes[key];
			if (value !== mandatoryValue)
				return false;
		}
		return true;
	},
	getWiredExport : function() {
		return this._wiredExport;
	},
	unwire : function() {
		this._wiredExport = null;
	}
};
